security/advancedtls/provider.go

```diff
@@ -0,0 +1,184 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package advancedtls
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"google.golang.org/grpc/credentials/tls/certprovider"
+)
+
+const (
+	defaultTestTimeout = 1 * time.Second
+)
+
+// IdentityPemFileProviderOptions contains fields to be filled out by users
+// for obtaining identity credential files.
+type IdentityPemFileProviderOptions struct {
+	CertFile string
+	KeyFile  string
+	Interval time.Duration
+}
+
+// IdentityPemFileProvider keeps identity credential implementations up to
+// date with secrets that they rely on to secure communications on the
+// underlying channel.
+type IdentityPemFileProvider struct {
+	distributor *certprovider.Distributor
+	closeFunc   func()
+}
+
+// RootPemFileProviderOptions contains fields to be filled out by users
+// for obtaining root credential files.
+type RootPemFileProviderOptions struct {
+	TrustFile string
+	Interval  time.Duration
+}
+
+// RootPemFileProvider keeps root credential implementations up to
+// date with secrets that they rely on to secure communications on the
+// underlying channel.
+type RootPemFileProvider struct {
+	distributor *certprovider.Distributor
+	closeFunc   func()
+}
+
+// NewIdentityPemFileProvider uses IdentityPemFileProviderOptions to
+// construct a IdentityPemFileProvider.
+func NewIdentityPemFileProvider(o *IdentityPemFileProviderOptions, quit chan bool) (*IdentityPemFileProvider, error) {
+	if o.CertFile == "" {
+		return nil, fmt.Errorf("users need to specify certificate file path in NewIdentityPemFileProvider")
+	}
+	if o.KeyFile == "" {
+		return nil, fmt.Errorf("users need to specify key file path in NewIdentityPemFileProvider")
+	}
+	// If interval is not set by users explicitly, set it to 1 hour by default.
+	if o.Interval == 0*time.Nanosecond {
+		o.Interval = 1 * time.Hour
+	}
+	provider := &IdentityPemFileProvider{}
+	// quit := make(chan bool)
+	provider.distributor = certprovider.NewDistributor()
+	// A goroutine to pull file changes.
+	go func(quit chan bool) {
+		for {
+			// Read identity certs from PEM files.
+			identityCert, err := o.ReadKeyAndCerts()
```

Comment:
The reading code is simple enough. How about we just expanding it here, instead of having another function? Same for the trust cert reading.

Reply:
Yeah, that sounds good.